src/FlowChart.hs

```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE RecordWildCards #-}
+
+import Algebra.Graph.Labelled as LG
+import Text.Megaparsec hiding (State)
+import Text.Megaparsec as M
+import Text.Megaparsec.Char (alphaNumChar, space1, string)
+import qualified Text.Megaparsec.Char.Lexer as L
+
+type Parser = Parsec Void Text
+
+data Orientation = TB | TD | BT | RL | LR deriving (Eq, Show)
+
+data Diagram
+  = FlowChart
+      { orientation :: Orientation,
+        graph :: [Graph (Style, Maybe Text) [(Text, Maybe (Bracket, Text))]]
```

Comment:
Some comments,

*   You want to use `Text` not `String`. See [https://free.cofree.io/2020/05/06/string-types/](https://free.cofree.io/2020/05/06/string-types/)
*   What does `Maybe String` refer to? And what does `Maybe (Bracket, String)` refer to? I think we can improve on these types, generally by creating new types. But it is hard to do that without understanding the _semantics_ of these types. You have it in your mind, but a casual reader - like me - would have trouble understanding the semantics. You can begin by writing comments explaining the semantics, and then the answer to how to refactor these into new types would naturally become evident.

Reply:
Also, it is generally a good idea to dedicate a separate type for the graph, something like:

```haskell
type FlowChartGraph = Graph Edge Vertex
```

Then you define `Edge` as a product type:

```haskell
data Edge = Edge 
   { edgeStyle :: Style
   , edgeFoo :: ..
```

As well `Vertex` (or `Node` of the flow chart) as:

```haskell
data Node = Node
  {  nodeLabel :: Text
  ,  ...
  }
```

The idea is this: just by reading the types we should be able to tell the semantics of the data as much as posible. So we 'design' types in a way that achieves that goal. So verbosity is completely fine (and being concise is an anti-goal).

Reply:
It is also not clear to me why `graph` above is a list rather than a single value. There is just *one* graph in a flow chart right?

Reply:
Likewise why vertex nodes are a list is not clear to me. Basically I'm curious why your type is not this:

```haskell
 graph :: Graph (LinkStyle, Maybe LinkText) (NodeId, Maybe (Bracket, NodeName))
```

(note the absence of square brackets)

Reply:
I am thinking of a way to represent something like this `a --> b & c--> d` 
the alternative way is to parse it as 
```
a ---> b
a ---> c
b ---> d
c ---> d
```
but I haven't thought of an easy way to achieve this, so at the moment i group `b` and `c` together as a list

Reply:
Your AST type doesn't necessarily have to map 1-on-1 to the syntax being parsed. You can construct arbitrary data in your parser code anyway.

You might want to consider using `edges` and `overlays` to basically 'merge' these edges in the single graph: https://hackage.haskell.org/package/algebraic-graphs-0.5/docs/Algebra-Graph-Labelled.html#v:overlay

Reply:
The "overlay" concept is used pretty commonly in this library. [This](https://blogs.ncl.ac.uk/andreymokhov/an-algebra-of-graphs/) is a good introduction. 

The idea here is that we 'composing' larger graphs from smaller ones. Even a single vertex is a graph (it is a graph with one vertex). Likewise a single edge is a graph with two vertices and one edge. You can combine them (`overlay`) to produce larger graphs. This is usually how one constructs graphs algebraically.

Reply:
I just overlayed the graphs into a single graph.

Reply:
Yea, this is an improvement, we are getting closer I think.

```
type FlowChartGraph = Graph Edge [Node]
```

Why `[Node]`? A vertex is a *single* node, not a list of nodes?

Also, as I understand, the `algebraic-graphs` requires the edge `e` to be a Monoid only to differentiate edges that are non-existent. `Edge "" Nothing` is hackish way to denote a non-existant edge. `Nothing` would be better, so you can use `Maybe Edge` as the edge type and drop your Monoid instance.

Reply:
I think the type shall be ok now